apps/definition/test/definition_test.exs

```diff
@@ -17,28 +18,47 @@ defmodule DefinitionTest do
     end
 
     def migrate(%__MODULE__{version: 1} = old) do
-      struct(__MODULE__, %{version: 2, bar: String.to_integer(old.bar)})
+      struct(__MODULE__, %{version: 2, id: old.id, bar: String.to_integer(old.bar)})
       |> Ok.ok()
     end
 
     defmodule V1 do
       use Definition.Schema
 
       def s do
-        schema(%Foo{version: spec(fn v -> v == 1 end), bar: spec(is_binary())})
+        schema(%Foo{
+          version: spec(fn v -> v == 1 end),
+          id: required_string(),
+          bar: spec(is_binary())
+        })
       end
     end
 
     defmodule V2 do
       use Definition.Schema
 
       def s do
-        schema(%Foo{version: spec(fn v -> v == 2 end), bar: spec(is_integer())})
+        schema(%Foo{
+          version: spec(fn v -> v == 2 end),
+          id: required_string(),
+          bar: spec(is_integer())
+        })
       end
     end
   end
 
   describe "__using__/1" do
+    setup do
+      fake_uuid = "fake_uuid"
+      allow(UUID.uuid4(), return: fake_uuid)
+      {:ok, [fake_uuid: fake_uuid]}
+    end
+
+    test "generates an id if not present on input", %{fake_uuid: fake_uuid} do
+      input = %{version: 2, bar: 9001}
+      assert {:ok, %Foo{version: 2, id: ^fake_uuid, bar: 9001, baz: nil}} = Foo.new(input)
+    end
+
```

Comment:
Do we need a test that it preserves id's when they are present?

Reply:
Yes and done